### PR TITLE
Create room api endpoint

### PR DESCRIPTION
**Pull request by JaimeSalas**

### CODE REVIEW

File: e2e/src/app.js [hunk 1]
```diff
@@ -1,38 +1,104 @@
...
+    document.getElementById('enrolmentsubmit')
+        .addEventListener('click', (evt) => {
+            evt.stopPropagation();
+            evt.preventDefault();
+            const room = selectedRoomValue('selectedroom');
+            const userId = userInputValue('userid');
+            canenrolRoom(
+                room,
+                userId,
+            ).then((canenrol) => {
```
Comment (Nasdan) [inline e2e/src/app.js]: canEnrol?

File: e2e/src/app.js [hunk 2]
```diff
@@ -1,38 +1,104 @@
...
+const handleEnrolment = (callback) => {
+    // NOTE: Check selected room.
+    // NOTE: Check input user.
+    document.getElementById('enrolmentsubmit')
+        .addEventListener('click', (evt) => {
+            evt.stopPropagation();
+            evt.preventDefault();
+            const room = selectedRoomValue('selectedroom');
+            const userId = userInputValue('userid');
+            canenrolRoom(
```
Comment (Nasdan) [inline e2e/src/app.js]: canEnrolRoom?

File: src/app.js [hunk 3]
```diff
@@ -1,14 +1,33 @@
 const express = require('express'),
   app = express(),
   bodyParser = require('body-parser'),
-  cors = require('cors')
-  random = require('./namespaces/random');
+  cors = require('cors'),
+  repository = require('./repositories/channelRepository'),
+  genericNamespace = require('./namespaces');
```
Comment (Nasdan) [inline src/app.js]: I don't understand about generic

File: src/app.js [hunk 4]
```diff
@@ -1,14 +1,33 @@
...
+app.use('/api/rooms', roomsRouter);
+
 const server = require('http').Server(app);
 const io = require('socket.io')(server);
-random(io);
-server.listen(3000);
\ No newline at end of file
+server.listen(3000);
+// TODO: Create env variable to initialize repository in memomory
+initiailizeChannels(io, channels, repository);
```
Comment (Nasdan) [inline src/app.js]: typo `initialize`

File: e2e/src/app.js (hunk 2 above)
Comment (brauliodiez) [reply e2e/src/app.js]: plus typo: canEnrollRoom I think

File: e2e/src/app.js [hunk 5]
```diff
@@ -1,38 +1,104 @@
...
+                room,
+                userId,
+            ).then((canenrol) => {
+                if (canenrol) {
```
Comment (brauliodiez) [inline e2e/src/app.js]: typo canEnroll

File: e2e/src/app.js [hunk 6]
```diff
@@ -1,38 +1,104 @@
...
+                    callback(null, room, userId);
+                } else {
+                    callback('Can not enrol to this channel');
```
Comment (brauliodiez) [inline e2e/src/app.js]: Typo enroll

File: e2e/src/app.js [hunk 7]
```diff
@@ -1,38 +1,104 @@
...
+        .addEventListener('click', (evt) => {
+            evt.stopPropagation();
+            const textElement = document.getElementById('message');
+            const message = _messageFactory.compose(textElement.value)
+            socket.emit('message', message);
+        });
+};
+
+document.addEventListener('DOMContentLoaded', () => {
+    let socket;
```
Comment (brauliodiez) [inline e2e/src/app.js]: Why do we need to define socket at this level? could it just be

Maybe is missing a return of the socket or something?

File: e2e/src/app.js [hunk 8]
```diff
@@ -1,38 +1,104 @@
...
+    handleEnrolment((err, room, userId) => {
+        if (!err) {
+            const socketParams = {
+                url: baseUrl,
+                channel: room,
+                options: {
+                    query: `user=${userId}`
+                },
+            };
+            socket = createSocket(socketParams);
```
Comment (brauliodiez) [inline e2e/src/app.js]:
```
const socket = createSocket(socketParams)  ?
```
Not sure if missing something with let socket

File: e2e/src/services/api.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+export const fetchRooms = () => (
+    fetch(`${apiUrl}/rooms`)
+        .then((response) => {
+            if (response.ok) {
+                return response.json();
+            }
+        })
+);
+
+export const canenrolRoom = (roomId, userId) => {
```
Comment (brauliodiez) [inline e2e/src/services/api.js]: canEnrollRoom ?

File: e2e/src/services/api.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    return fetch(`${apiUrl}/rooms/canenrol/${roomId}/user`, {
```
Comment (brauliodiez) [inline e2e/src/services/api.js]: Same thing here, url typo
```
canenroll
```

File: e2e/src/services/uiHelpers.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+export const populateRooms = (rooms, elementId) => {
+    const roomsContainer = document.getElementById(elementId);
+    const roomItems = rooms.map((r) => {
+        const roomItem = document.createElement('li');
+        roomItem.innerText = r;
+        return roomItem;
+    });
+    roomItems.forEach((ri) => roomsContainer.appendChild(ri));
+};
```
Comment (brauliodiez) [inline e2e/src/services/uiHelpers.js]: Just minor feedback, Proposal:

```javascript
export const populateRooms = (rooms, elementId) => {
    const roomsContainer = document.getElementById(elementId);
    const roomItems = rooms.map((r) => 
            populateRoom(r);
   );
    roomItems.forEach((ri) => roomsContainer.appendChild(ri));
};

const populateRoom = (roomName) => {
  const roomItem = document.createElement('li');
  roomItem.innerText = r;

  return roomItem;
}
```

File: src/routes/rooms.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        .get('/', async (_, res, next) => {
+            try {
+                const rooms = await repository.channelList();
+                console.log(rooms);
+                res.json(rooms);
+            } catch (error) {
+                next(error);
+            }
+        })
+        .post('/canenrol/:roomId/user', async (req, res, next) => {
```
Comment (brauliodiez) [inline src/routes/rooms.js]: typo

canenroll

**Comment by brauliodiez** [comment]
Good stuff, just found some typos that would need to be corrected (specially the ones the affect to rest-api url's, those ones will be used on the training.

**Comment by brauliodiez** [comment]
Just realized rooms list is returning only an array of room names, should we rather return id and name?

something like

```javascript
[
  {
    id: 1,
    name: 'random'
}, 
{
   id:2,
  name: 'pull requests'
}
]
```

I think having an ID could simulate better a chat list (e.g. you can change the name of a chat room, and should affect the internal id), what do you think?

**Comment by JaimeSalas** [comment]
There is no id for rooms because name matches with the generated socket.io channels, this is used for websocket routing. A bit a refactor it's need it to change it, but it's not a big deal. What about slack when you generate a channel can you change the name later? Bear on mind that room name and roomId, matches as well because it got not sense to have two channels with the same name at this level (maybe it's something to get on mind), although a server validation to not duplicate channels will be need it. So I'm completely open to use and id, but I think that has to match with the socket routing, so maybe a number it's not the best option.

**Comment by brauliodiez** [comment]
I was thinking about having ID's, as you mentioned the guid solution makes a lot of sense to me, but it can be ok to have fixed rooms names if needed

File: src/app.js (hunk 4 above)
Comment (JaimeSalas) [reply src/app.js]: Thanks!!

File: e2e/src/app.js (hunk 7 above)
Comment (JaimeSalas) [reply e2e/src/app.js]: No need at last, because I am using callbacks at last.

File: e2e/src/app.js (hunk 8 above)
Comment (JaimeSalas) [reply e2e/src/app.js]: Related with prior comment.

**Pull request merged**